sdks/java/io/rabbitmq/src/test/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIOTest.java

```diff
@@ -0,0 +1,315 @@
+    final int maxNumRecords = 10;
+    PCollection<RabbitMqMessage> raw =
+        p.apply(
+            RabbitMqIO.read()
+                .withUri("amqp://guest:guest@localhost:" + port)
+                .withQueue("READ")
+                .withMaxNumRecords(maxNumRecords));
+    PCollection<String> output =
+        raw.apply(
+            MapElements.into(TypeDescriptors.strings())
+                .via(
+                    (RabbitMqMessage message) ->
+                        new String(message.getBody(), StandardCharsets.UTF_8)));
+
+    List<String> records =
+        generateRecords(maxNumRecords)
+            .stream()
+            .map(record -> new String(record, StandardCharsets.UTF_8))
+            .collect(Collectors.toList());
+    PAssert.that(output).containsInAnyOrder(records);
+
+    ConnectionFactory connectionFactory = new ConnectionFactory();
+    connectionFactory.setUri("amqp://guest:guest@localhost:" + port);
+    Connection connection = null;
+    Channel channel = null;
+    try {
+      connection = connectionFactory.newConnection();
+      channel = connection.createChannel();
+      channel.queueDeclare("READ", false, false, false, null);
+      for (String record : records) {
+        channel.basicPublish("", "READ", null, record.getBytes(StandardCharsets.UTF_8));
+      }
+
+      p.run();
+    } finally {
+      if (channel != null) {
+        channel.close();
+      }
+      if (connection != null) {
+        connection.close();
+      }
+    }
+  }
+
+  @Test(timeout = 60 * 1000)
+  public void testReadExchange() throws Exception {
+    final int maxNumRecords = 10;
+    PCollection<RabbitMqMessage> raw =
+        p.apply(
+            RabbitMqIO.read()
+                .withUri("amqp://guest:guest@localhost:" + port)
+                .withExchange("READEXCHANGE", "fanout", "test")
+                .withMaxNumRecords(maxNumRecords));
+    PCollection<String> output =
+        raw.apply(
+            MapElements.into(TypeDescriptors.strings())
+                .via(
+                    (RabbitMqMessage message) ->
+                        new String(message.getBody(), StandardCharsets.UTF_8)));
+
+    List<String> records =
+        generateRecords(maxNumRecords)
+            .stream()
+            .map(record -> new String(record, StandardCharsets.UTF_8))
+            .collect(Collectors.toList());
+    PAssert.that(output).containsInAnyOrder(records);
+
+    ConnectionFactory connectionFactory = new ConnectionFactory();
+    connectionFactory.setUri("amqp://guest:guest@localhost:" + port);
+    Connection connection = null;
+    Channel channel = null;
+    try {
+      connection = connectionFactory.newConnection();
+      channel = connection.createChannel();
+      channel.exchangeDeclare("READEXCHANGE", "fanout");
+      Channel finalChannel = channel;
+      Thread publisher =
+          new Thread(
+              () -> {
+                try {
+                  Thread.sleep(5000);
```

Comment:
I think you missed that one: In a previous comment I suggested to replace this absolute value sleep by something mode dynamical to avoid flakiness on the tests in case of load of the jenkins server. I guess this sleep is to let the broker enough time to create the exchange. Now, I don't think the sleep is actually needed because exchangeDeclare blocks until it gets a response from the server and throws an IOException otherwise. IMHO it would have been needed if you would have used exchangeDeclareNoWait